dateutil/tz/_factories.py

```diff
@@ -11,6 +13,29 @@ def __call__(cls):
             cls.__instance = super(_TzSingleton, cls).__call__()
         return cls.__instance
 
+
+class _TzChangeSingleton(type):
+
+    """
+    A :class: `tzlocalsingleton`
+               Singleton that guards against TZ changes
+    """
+
+    def __init__(cls, *args, **kwargs):
+        cls.__instance = None
+        cls.__tz = time.tzname
```

Comment:
This guard is not as robust as one would hope, though it's probably a pretty good sentinel, consider:

```python
with TZEnvContext('EST5EDT'):
    l1 = tz.tzlocal()

with TZEnvContext('EST6EDT'):
    l2 = tz.tzlocal()
```